### PR TITLE
Add Flash Attention Layer

### PR DESCRIPTION
**Pull request by tchaton**
This PR adds Flash Attention from https://github.com/HazyResearch/flash-attention.

LDM + Flash Attention works on T4 :fire:
However, inference time is 8.698secs for batch 1 on T4. Quite higher than I expected compared to A10.
On 3090 RTX, here are the numbers I am getting:
* DeepSpeed Triton Flash Attention: 1.545 sec
* [Original Flash Attention](https://github.com/HazyResearch/flash-attention): 1.491 sec (slightly faster)

image

### CODE REVIEW

File: ldm/deepspeed_replace.py [hunk 1]
```diff
@@ -271,37 +273,164 @@ def apply(self, module, enable_cuda_graph=True):
...
+
+    def forward(self, input, context=None, input_mask=None):
+        from flash_attn.flash_attn_interface import _flash_attn_forward
+
+        batch_size = input.shape[0]
+        seqlen = input.shape[1]
+        cu_seqlens = torch.arange(0, (batch_size + 1) * seqlen, step=seqlen, dtype=torch.int32, device=self.attn_kw.device)
+        do_flash_attn = (input.shape[-1] / self.heads <= 128) and context is None
+
+        if not do_flash_attn:
```
Comment (lantiga) [inline ldm/deepspeed_replace.py:355]: Shouldn't we output some kind of warning here, if the expectation is that this is running with FlashAttention? It kind of breaks the contract.

**Review by lantiga — APPROVED** [review]
Looks great!

File: ldm/deepspeed_replace.py (hunk 1 above)
Comment (tchaton) [reply ldm/deepspeed_replace.py:355]: I followed Flash Attention & DeepSpeed internal contract. I didn't change the logic. Unfortunately, flash attention isn't implemented for context and heads higher than 128.

So problably, there are extra speed gains to get there by implementing the kernels.

Comment (lantiga) [reply ldm/deepspeed_replace.py:355]: I see, we should make this explicit in the docstrings at the very least, or even better by printing a warning the first time the condition fails.

Comment (tchaton) [reply ldm/deepspeed_replace.py:355]: Done !

**Pull request merged**